### ISSUE TITLE
How does masking affect forward and back propagation through time in LSTMs?

### ISSUE THREAD

**Issue by jaweriaamjad** [keras-team/tf-keras#179]
I am using Tensorflow API to code variational autoencoder for time series data. My model is as follows:

```def get_model(max_ts, padding_value):
    # encoder
    inp_x = tfk.layers.Input(shape=(max_ts, 1))
    masked_x = tfk.layers.Masking(mask_value=padding_value)(inp_x) 
    x = tfk.layers.LSTM(64,return_sequences=True)(masked_x)
    x = tfk.layers.LSTM(64)(x)
    z = tfk.layers.Dense(32, activation="relu")(x)
    z_mean = tfk.layers.Dense(latent_dim)(z)
    z_log_sigma = tfk.layers.Dense(latent_dim)(z)
    encoder = tfk.Model(inp_x, [z_mean, z_log_sigma])
    
    # decoder 
    inp_z = tfk.layers.Input(shape=(latent_dim,))
    y = masked_repeatvector(max_ts, padding_value)([inp_z, inp_x])
    y = tfk.layers.LSTM(64, return_sequences=True)(y)
    out = tfk.layers.TimeDistributed(tfk.layers.Dense(1))(y)
    decoder = tfk.Model(inp_z, out)   
    
    # encoder & decoder 
    z_mean, z_log_sigma = encoder(inp_x)
    z = tfk.layers.Lambda(sampling)([z_mean, z_log_sigma])
    out = decoder(z)
    vae = tfk.Model(inp_x, out)
    vae.add_loss(vae_loss(inp_x, out, max_ts, z_log_sigma, z_mean))
    # vae.add_loss(masked_vae_loss(inp_x, out, max_ts, z_log_sigma, z_mean, padding_value))
    return vae, encoder, decoder
```

As you can see my data is variable length so naturally I am using padding at the input. however, the padding information is lost at the end of the decoder. So to ensure the decoder accounts for the padding values, I have two options 1) either include a masking layer after the repeatvector layer 2) include a masked loss. I have gone with the second solution. My loss function is as follow:

```
def masked_vae_loss(original, out, ts,  z_log_sigma, z_mean, padding_value):
    mask = tf.cast(tf.not_equal(original, padding_value), dtype=tf.float32)
    reconstruction = K.mean(K.square(tf.math.multiply(original - out,mask)))*ts
    kl = -0.5 * K.mean(1 + z_log_sigma - K.square(z_mean) - K.exp(z_log_sigma))
    return reconstruction + kl
``` 

My model works but I get outputs with last value repeated for the padded values. Same thing happens when i use the decoder to generate timeseries data from sampled latents. I am having trouble understanding the outputs. From my research, I know that in Tensorflow, LSTM has been implemented so that last hidden states is copied for the padded/masked values. Does it also happen when the loss is maksed and no explicit masking is provided?
